tensorflow/core/kernels/non_max_suppression_op.cc

```diff
@@ -195,6 +222,229 @@ void DoNonMaxSuppressionOp(
+  std::vector<float> clip_window{0, 0, 1, 1};
+  
+  // [num_batches, per_batch_size, 4]
+  std::vector<std::vector<float>> nmsed_boxes(num_batches);
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<float>> nmsed_scores(num_batches);
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<float>> nmsed_classes(num_batches);
+  // [num_batches]
+  std::vector<int> final_valid_detections;
+  // [num_batches, per_batch_size]
+  std::vector<std::vector<int>> selected_indices(num_batches);
+
+  int per_batch_size = total_size_per_batch;
+
+  //perform non_max_suppression operation for each batch independently
+  for (int batch = 0; batch < num_batches; ++batch) {
+    // dims of per_batch_boxes [num_boxes, q, 4]
+    Tensor per_batch_boxes = inp_boxes.Slice(batch, batch+1);
+    // dims of per_batch_scores [num_boxes, num_classes]
+    Tensor per_batch_scores = inp_scores.Slice(batch, batch+1);
+
+    struct ResultCandidate {
+      int box_index;
+      float score;
+      int class_idx;
+      float box_coord[4];
+    };
+
+    auto rc_cmp = [](const ResultCandidate rc_i, const ResultCandidate rc_j) {
+        return rc_i.score < rc_j.score;
+    };
+    std::priority_queue<ResultCandidate, std::vector<ResultCandidate>, 
+      decltype(rc_cmp)> result_candidate_pq(rc_cmp);
+
+    float * scoresData = per_batch_scores.unaligned_flat<float>().data();
+    float * boxesData = per_batch_boxes.unaligned_flat<float>().data();
+
+    //Iterate through all classes
+    for (int class_idx = 0; class_idx < num_classes; ++class_idx) {
+      std::vector<float> scores_data;
+      std::vector<float> boxes_data_vec;
+
+      for (int box = 0; box < num_boxes; ++box) {
+        //Get the scores per class 
+        //scores_data dim is [num_boxes].
+        scores_data.push_back(scoresData[box * num_classes + class_idx]);
+        for(int cid = 0; cid < 4; ++cid){
+          if(q > 1){
+            //Get the boxes per class. boxes_data_vec dims is [num_boxes, 4]
+            boxes_data_vec.push_back(boxesData[box * q * 4 + 
+                  class_idx * 4 + cid]);
+          }
+          else {
+              boxes_data_vec.push_back(boxesData[box * 4 + cid]);
+          }
+        }
+      }
+        
+      //Copy boxes_data_vec to a tensor
+      TensorShape boxesShape({num_boxes, 4});
+      Tensor boxes(per_batch_boxes.dtype(), boxesShape);
+      std::copy_n(boxes_data_vec.begin(), boxes_data_vec.size(), 
+          boxes.unaligned_flat<float>().data());
+
+      const int size_per_class = std::min(max_size_per_class, num_boxes);
+      // Do NMS, get the candidate indices of form vector<int>
+      // Data structure for selection candidate in NMS.
+      struct Candidate {
+        int box_index;
+        float score;
+      };
+      auto cmp = [](const Candidate bs_i, const Candidate bs_j) {
+        return bs_i.score < bs_j.score;
+      };
+      std::priority_queue<Candidate, std::deque<Candidate>, decltype(cmp)>
```

Comment:
why do you need a priority queue (for  candidate_priority_queue and result_candidate_pq) vector/queue should work - you just need to sort them once.

Reply:
I followed the same technique as used by the existing DoNonMaxSuppressionOp() function. It uses a priority queue based on scores. Do you want me to change these?

Reply:
Updated to use vector and sorted it